datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -238,8 +238,17 @@ impl CommonSubexprEliminate {
         let rewritten = pop_expr(&mut rewritten)?;
 
         if affected_id.is_empty() {
+            // Alias aggregation epxressions if they have changed
+            // TODO: This should have really been identified above and handled in the `else` branch
+            let aggr_exprs = new_aggr_expr
+                .iter()
+                .zip(aggr_expr.iter())
+                .map(|(new_expr, old_expr)| {
+                    new_expr.clone().alias_if_changed(old_expr.display_name()?)
+                })
+                .collect::<Result<Vec<Expr>>>()?;
             // Since group_epxr changes, schema changes also. Use try_new method.
-            Aggregate::try_new(Arc::new(new_input), new_group_expr, new_aggr_expr)
+            Aggregate::try_new(Arc::new(new_input), new_group_expr, aggr_exprs)
                 .map(LogicalPlan::Aggregate)
```

Comment:
Ok, I think I understand what is going on:
- in general the root aggregation expressions need not be rewritten/aliased above since they can only appear once (i.e. they're not common), so this is ok
- however, some of their child expressions may get rewritten; an example is the query `select distinct on (c1 + 1) c2 from test order by c1 + 1, c3`, where the initial aggregation of the sole projected field gets rewritten from `FIRST_VALUE(test.column2) ORDER BY [test.column1 + Int64(1) ASC NULLS LAST, test.column3 ASC NULLS LAST]` to `FIRST_VALUE(test.column2) ORDER BY [test.column1 + Int64(1)Int64(1)test.column1 AS test.column1 + Int64(1) ASC NULLS LAST, test.column3 ASC NULLS LAST]` (i.e. `c1 + 1` is a common expression of it and the `ON` expression so it gets rewritten)
- it seems this was hidden before thanks to using `Aggregate::try_new_with_schema` with the original schema, however with #7919 this is no longer the case, so the newly created schema is mismatched with schema expected by the parent node

I think this explains the behavior I've encountered here, and that this behavior is not anomalous. The only question is whether there is a better way to align the new aggregation schema with the old one besides aliasing as above.

Reply:
Maybe @waynexia has some thoughts

Reply:
Btw, I've now clarified this comment since it was misleading, i.e. the aliasing should really be done in `build_recover_project_plan` once qualified aliasing is enabled.